<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Examples · BioRecordsProcessing.jl</title><meta name="title" content="Examples · BioRecordsProcessing.jl"/><meta property="og:title" content="Examples · BioRecordsProcessing.jl"/><meta property="twitter:title" content="Examples · BioRecordsProcessing.jl"/><meta name="description" content="Documentation for BioRecordsProcessing.jl."/><meta property="og:description" content="Documentation for BioRecordsProcessing.jl."/><meta property="twitter:description" content="Documentation for BioRecordsProcessing.jl."/><meta property="og:url" content="https://jonathanBieler.github.io/BioRecordsProcessing.jl/examples/"/><meta property="twitter:url" content="https://jonathanBieler.github.io/BioRecordsProcessing.jl/examples/"/><link rel="canonical" href="https://jonathanBieler.github.io/BioRecordsProcessing.jl/examples/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">BioRecordsProcessing.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Manual</a></li><li class="is-active"><a class="tocitem" href>Examples</a><ul class="internal"><li><a class="tocitem" href="#Reading-a-FASTA-file-into-memory"><span>Reading a FASTA file into memory</span></a></li><li><a class="tocitem" href="#Transforming-a-FASTA-file"><span>Transforming a FASTA file</span></a></li><li><a class="tocitem" href="#Reading-a-pair-of-FASTA-file"><span>Reading a pair of FASTA file</span></a></li><li><a class="tocitem" href="#Processing-all-files-in-a-directory"><span>Processing all files in a directory</span></a></li><li><a class="tocitem" href="#Write-sequences-in-memory-into-a-file"><span>Write sequences in memory into a file</span></a></li><li class="toplevel"><a class="tocitem" href="#Cookbook"><span>Cookbook</span></a></li><li><a class="tocitem" href="#Loading-paired-end-reads-from-a-BAM-file-in-a-genomic-interval"><span>Loading paired-end reads from a BAM file in a genomic interval</span></a></li><li><a class="tocitem" href="#Generate-artificial-FASTQ-reads"><span>Generate artificial FASTQ reads</span></a></li><li><a class="tocitem" href="#Reading-a-VCF-in-a-DataFrame"><span>Reading a VCF in a DataFrame</span></a></li><li><a class="tocitem" href="#Aligning-a-FASTQ-file-to-the-reference-genome-:"><span>Aligning a FASTQ file to the reference genome :</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Examples</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Examples</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/jonathanBieler/BioRecordsProcessing.jl/blob/main/docs/src/examples.md#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Examples"><a class="docs-heading-anchor" href="#Examples">Examples</a><a id="Examples-1"></a><a class="docs-heading-anchor-permalink" href="#Examples" title="Permalink"></a></h1><p>All examples use <code>FASTA</code> files but should work with <code>FASTQ</code>, compressed <code>.gz</code> files, <code>VCF</code> from  <code>VariantCallFormat.jl</code> and <code>XAM.jl</code> types. More examples can be found in the tests.</p><h2 id="Reading-a-FASTA-file-into-memory"><a class="docs-heading-anchor" href="#Reading-a-FASTA-file-into-memory">Reading a FASTA file into memory</a><a id="Reading-a-FASTA-file-into-memory-1"></a><a class="docs-heading-anchor-permalink" href="#Reading-a-FASTA-file-into-memory" title="Permalink"></a></h2><pre><code class="language-julia hljs">using BioRecordsProcessing, FASTX, BioSequences

# the file contains two 20bp reads
p = Pipeline(
    Reader(FASTX.FASTA, File(filepath)),
    record -&gt; begin
        sequence(LongDNA{4}, record)
    end,
    Collect(LongDNA{4}),
)
run(p)

# output
2-element Vector{LongSequence{DNAAlphabet{4}}}:
 CTTGGCATACTCAAACTCTT
 TGGCATACTCACTAACTCTT</code></pre><h2 id="Transforming-a-FASTA-file"><a class="docs-heading-anchor" href="#Transforming-a-FASTA-file">Transforming a FASTA file</a><a id="Transforming-a-FASTA-file-1"></a><a class="docs-heading-anchor-permalink" href="#Transforming-a-FASTA-file" title="Permalink"></a></h2><pre><code class="language-julia hljs">using BioRecordsProcessing, FASTX, BioSequences

# the file contains two 20bp reads, trim first 10bp
p = Pipeline(
    Reader(FASTX.FASTA, File(filepath)),
    record -&gt; begin
        seq = sequence(LongDNA{4}, record)
        FASTA.Record(FASTA.identifier(record), seq[10:end])
    end,
    Writer(FASTX.FASTA, dir; suffix = &quot;.trimmed&quot;),
)
out = run(p)
run(`head $out`)

# output
&gt;seq1
CTCAAACTCTT
&gt;seq2
CACTAACTCTT
Process(`head /var/folders/8g/xj7pzy251n53px06l17vr0_00000gr/T/jl_mL4pM7/test_1.trimmed.fa`, ProcessExited(0))</code></pre><h2 id="Reading-a-pair-of-FASTA-file"><a class="docs-heading-anchor" href="#Reading-a-pair-of-FASTA-file">Reading a pair of FASTA file</a><a id="Reading-a-pair-of-FASTA-file-1"></a><a class="docs-heading-anchor-permalink" href="#Reading-a-pair-of-FASTA-file" title="Permalink"></a></h2><pre><code class="language-julia hljs">using BioRecordsProcessing, FASTX, BioSequences

# first in pair is named &quot;_1.fasta&quot;, second &quot;_2.fasta&quot;
p = Pipeline(
    Reader(FASTX.FASTA, File(filepath; second_in_pair = x -&gt; replace(x, &quot;_1&quot; =&gt; &quot;_2&quot;))),
    (r1, r2) -&gt; begin
        sequence(LongDNA{4}, r1), sequence(LongDNA{4}, r2)
    end,
    Collect(LongDNA{4}; paired = true),
)
run(p)

# output
2-element Vector{Tuple{LongSequence{DNAAlphabet{4}}, LongSequence{DNAAlphabet{4}}}}:
 (CTTGGCATACTCAAACTCTT, GCAAACTCTTCTTGGCATAC)
 (TGGCATACTCACTAACTCTT, ATACTCAAACTCTTCTTGGC)</code></pre><h2 id="Processing-all-files-in-a-directory"><a class="docs-heading-anchor" href="#Processing-all-files-in-a-directory">Processing all files in a directory</a><a id="Processing-all-files-in-a-directory-1"></a><a class="docs-heading-anchor-permalink" href="#Processing-all-files-in-a-directory" title="Permalink"></a></h2><pre><code class="language-julia hljs">using BioRecordsProcessing, FASTX, BioSequences


p = Pipeline(
    Reader(FASTX.FASTA, Directory(dir, &quot;*.fa&quot;)),
    record -&gt; begin
        seq = sequence(LongDNA{4}, record)
        FASTA.Record(FASTA.identifier(record), seq[10:end])
    end,
    Writer(FASTX.FASTA, dir; suffix = &quot;.trimmed&quot;),
)
out = run(p; verbose = false)
basename.(out)# run returns the path to output files

# output
2-element Vector{String}:
 &quot;test_1.trimmed.fa&quot;
 &quot;test_2.trimmed.fa&quot;</code></pre><h2 id="Write-sequences-in-memory-into-a-file"><a class="docs-heading-anchor" href="#Write-sequences-in-memory-into-a-file">Write sequences in memory into a file</a><a id="Write-sequences-in-memory-into-a-file-1"></a><a class="docs-heading-anchor-permalink" href="#Write-sequences-in-memory-into-a-file" title="Permalink"></a></h2><pre><code class="language-julia hljs">using BioRecordsProcessing, FASTX, BioSequences

data = [FASTA.Record(&quot;seq1&quot;, dna&quot;ATGC&quot;)]

p = Pipeline(
    Buffer(data; filename = &quot;test.fa&quot;),
    Writer(FASTX.FASTA, dir),
)
out = run(p)
run(`head $out`)

# output
&gt;seq1
ATGC
Process(`head /var/folders/8g/xj7pzy251n53px06l17vr0_00000gr/T/jl_NSdfEq/test.fa`, ProcessExited(0))</code></pre><h1 id="Cookbook"><a class="docs-heading-anchor" href="#Cookbook">Cookbook</a><a id="Cookbook-1"></a><a class="docs-heading-anchor-permalink" href="#Cookbook" title="Permalink"></a></h1><h2 id="Loading-paired-end-reads-from-a-BAM-file-in-a-genomic-interval"><a class="docs-heading-anchor" href="#Loading-paired-end-reads-from-a-BAM-file-in-a-genomic-interval">Loading paired-end reads from a BAM file in a genomic interval</a><a id="Loading-paired-end-reads-from-a-BAM-file-in-a-genomic-interval-1"></a><a class="docs-heading-anchor-permalink" href="#Loading-paired-end-reads-from-a-BAM-file-in-a-genomic-interval" title="Permalink"></a></h2><p>In general records can be grouped using a <code>RecordGrouper</code>. For pair-end BAMs <code>BAMPairedReadGrouper</code> is provided. It will group the two first reads (primary alignments only) with the same read name it encounters and release them for processing as a pair:</p><pre><code class="language-julia hljs">using BioRecordsProcessing, XAM, GenomicFeatures

region = Interval(&quot;9&quot;, 22331023, 24542023)

p = Pipeline(
    Reader(XAM.BAM, File(bamfile; interval = region)),
    BioRecordsProcessing.BAMPairedReadGrouper(),
    (r1,r2) -&gt; begin 
        (r1,r2)
    end,
    Collect(Tuple{XAM.BAM.Record, XAM.BAM.Record}),
)
out = run(p)</code></pre><h2 id="Generate-artificial-FASTQ-reads"><a class="docs-heading-anchor" href="#Generate-artificial-FASTQ-reads">Generate artificial FASTQ reads</a><a id="Generate-artificial-FASTQ-reads-1"></a><a class="docs-heading-anchor-permalink" href="#Generate-artificial-FASTQ-reads" title="Permalink"></a></h2><p>A generator can be passed as input in a Buffer : </p><pre><code class="language-julia hljs">using BioRecordsProcessing, FASTX, BioSequences

generate_read(i) = FASTA.Record(&quot;seq$i&quot;, randdnaseq(150))

p = Pipeline(
    Buffer(generate_read(i) for i in 1:10; filename = &quot;test.fa&quot;),
    Writer(FASTX.FASTA, dir),
)
out = run(p)

# output
&quot;/tmp/jl_0mSqQJ/test.fastq&quot;</code></pre><p>Paired-end reads can also be generated :</p><pre><code class="language-julia hljs">generate_read(i) = FASTQ.Record(&quot;seq$i&quot;, randdnaseq(150), fill(UInt8(40), 150))
generate_reads(i) = (generate_read(i), generate_read(i))

p = Pipeline(
    Buffer(generate_reads(i) for i in 1:10; filename = &quot;test_R1.fastq.gz&quot;),
    Writer(FASTX.FASTQ, dir; paired = true, second_in_pair = x -&gt; replace(x, &quot;_R1&quot; =&gt; &quot;_R2&quot;)),
)
out = run(p)

# output
2-element Vector{String}:
 &quot;/tmp/jl_srnqiA/test_R1.fastq.gz&quot;
 &quot;/tmp/jl_srnqiA/test_R2.fastq.gz&quot;</code></pre><h2 id="Reading-a-VCF-in-a-DataFrame"><a class="docs-heading-anchor" href="#Reading-a-VCF-in-a-DataFrame">Reading a VCF in a DataFrame</a><a id="Reading-a-VCF-in-a-DataFrame-1"></a><a class="docs-heading-anchor-permalink" href="#Reading-a-VCF-in-a-DataFrame" title="Permalink"></a></h2><p>Note : This probably wont&#39; work because of compatibilty issues, see : https://github.com/rasmushenningsson/VariantCallFormat.jl/issues/5</p><pre><code class="language-julia hljs">using VariantCallFormat

get_depth(r) = parse(Int, VCF.genotype(r, 1, &quot;DP&quot;))
get_vaf(r) = parse(Float64, VCF.genotype(r, 1, &quot;VAF&quot;))

T = NamedTuple{(:chr, :pos, :ref, :alt, :depth, :vaf, :quality), Tuple{String, Int64, String, String, Int64, String, Float64}}

p = Pipeline(
    Reader(VCF, File(filepath)),
    r -&gt; begin
        VCF.filter(r) !=  [&quot;PASS&quot;] &amp;&amp; return nothing #filter variants that are not PASS

        (chr = VCF.chrom(r), pos = VCF.pos(r), ref = VCF.ref(r), alt = VCF.alt(r)[1], 
        depth=get_depth(r), vaf=VCF.genotype(r, 1, &quot;VAF&quot;), quality = VCF.qual(r)
        )
    end,
    Collect(T),
)

df = run(p) |&gt; DataFrame</code></pre><h2 id="Aligning-a-FASTQ-file-to-the-reference-genome-:"><a class="docs-heading-anchor" href="#Aligning-a-FASTQ-file-to-the-reference-genome-:">Aligning a FASTQ file to the reference genome :</a><a id="Aligning-a-FASTQ-file-to-the-reference-genome-:-1"></a><a class="docs-heading-anchor-permalink" href="#Aligning-a-FASTQ-file-to-the-reference-genome-:" title="Permalink"></a></h2><p>See this blog post : https://jonathanbieler.github.io/blog/fastq2cnv/</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Manual</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.7.0 on <span class="colophon-date" title="Thursday 10 October 2024 15:28">Thursday 10 October 2024</span>. Using Julia version 1.7.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
