<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Manual · BioRecordsProcessing.jl</title><script data-outdated-warner src="assets/warner.js"></script><link rel="canonical" href="https://jonathanBieler.github.io/BioRecordsProcessing.jl/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>BioRecordsProcessing.jl</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Manual</a><ul class="internal"><li><a class="tocitem" href="#Sources"><span>Sources</span></a></li><li><a class="tocitem" href="#Sinks"><span>Sinks</span></a></li></ul></li><li><a class="tocitem" href="examples/">Examples</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Manual</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Manual</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/jonathanBieler/BioRecordsProcessing.jl/blob/main/docs/src/index.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="BioRecordsProcessing"><a class="docs-heading-anchor" href="#BioRecordsProcessing">BioRecordsProcessing</a><a id="BioRecordsProcessing-1"></a><a class="docs-heading-anchor-permalink" href="#BioRecordsProcessing" title="Permalink"></a></h1><p>In BioRecordsProcessing records are processed using a <code>Pipeline</code> that is constructed by taking a source (producing records), a user-defined function to process the records and a sink that will store the output of the processing function. The pipeline can then be run.</p><p>In this example a FASTA file is read from the disk, the sequence is extracted from the records and collected in an array :</p><pre><code class="language-julia hljs">using BioRecordsProcessing, FASTX, BioSequences

p = Pipeline(
    Reader(FASTX.FASTA, File(filepath)),
    record -&gt; begin
        sequence(LongDNA{4}, record)
    end,
    Collect(LongDNA{4}),
)
run(p)

# output
2-element Vector{LongSequence{DNAAlphabet{4}}}:
 CTTGGCATACTCAAACTCTT
 CTTGGCATACTCAAACTCTT</code></pre><p>By using different combinations of source and sink, and with user defined processing function, this allows to handle many common cases of biological records processing.</p><h3 id="Conventions"><a class="docs-heading-anchor" href="#Conventions">Conventions</a><a id="Conventions-1"></a><a class="docs-heading-anchor-permalink" href="#Conventions" title="Permalink"></a></h3><ul><li>If the processing function returns <code>nothing</code> the record will not be written to the sink, allowing to filter out records.</li><li>When writing a file to the disk the sink will get the filename from the source, so a source need to have a filename provided in this case. </li></ul><h2 id="Sources"><a class="docs-heading-anchor" href="#Sources">Sources</a><a id="Sources-1"></a><a class="docs-heading-anchor-permalink" href="#Sources" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="BioRecordsProcessing.Reader" href="#BioRecordsProcessing.Reader"><code>BioRecordsProcessing.Reader</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">Reader(record_module::Module, file_provider::F) where {F &lt;: AbstractFileProvider}</code></pre><p>Read a file or a directory on the disk and produce records of type <code>record_module.Record</code>.  The second argument can be a <code>File</code> or a <code>Directory</code>.</p><p>If a string is passed the second argment will default to <code>File</code>.</p><pre><code class="language- hljs">Reader(FASTX.FASTA, &quot;test.fa&quot;)
Reader(FASTX.FASTA, File(&quot;test.fa&quot;))
Reader(FASTX.FASTQ, Directory(&quot;data/&quot;, &quot;*.fastq&quot;))</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jonathanBieler/BioRecordsProcessing.jl/blob/03dd05429a737063e83a6e8c225d3c64c19526ed/src/Source.jl#L55-L70">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BioRecordsProcessing.Buffer" href="#BioRecordsProcessing.Buffer"><code>BioRecordsProcessing.Buffer</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">Buffer(data::Vector{T}; filename = &quot;&quot;)</code></pre><p>Use the array <code>data</code> as a source of records. An optional filename can be provided when a <code>Writer</code> is used as a source.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jonathanBieler/BioRecordsProcessing.jl/blob/03dd05429a737063e83a6e8c225d3c64c19526ed/src/Source.jl#L105-L112">source</a></section></article><h3 id="File-Providers"><a class="docs-heading-anchor" href="#File-Providers">File Providers</a><a id="File-Providers-1"></a><a class="docs-heading-anchor-permalink" href="#File-Providers" title="Permalink"></a></h3><p>Reader can take one of these files provider as agument :</p><article class="docstring"><header><a class="docstring-binding" id="BioRecordsProcessing.File" href="#BioRecordsProcessing.File"><code>BioRecordsProcessing.File</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">File(filename; second_in_pair = nothing)</code></pre><p>For paired files a function taking as argument the filename of the first file in pair and returning the filename of the second file can be provided. For example one can use <code>replace</code>  or a dictionnary, e.g. <code>second_in_pair = f1 -&gt; replace(f1, &quot;_1&quot; =&gt; &quot;_2&quot;)</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jonathanBieler/BioRecordsProcessing.jl/blob/03dd05429a737063e83a6e8c225d3c64c19526ed/src/Source.jl#L7-L15">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BioRecordsProcessing.Directory" href="#BioRecordsProcessing.Directory"><code>BioRecordsProcessing.Directory</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">Directory(directory::String, glob_pattern::String; second_in_pair = nothing)</code></pre><p>List all files matching the <code>glob_pattern</code> (See <a href="https://github.com/vtjnash/Glob.jl">Glob.jl</a>) in <code>directory</code>. For paired files a function taking as argument the filename of the first file in pair and returning the filename of the second file can be provided.</p><pre><code class="language- hljs">Directory(input_directory, &quot;*.fastq&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jonathanBieler/BioRecordsProcessing.jl/blob/03dd05429a737063e83a6e8c225d3c64c19526ed/src/Source.jl#L24-L36">source</a></section></article><h2 id="Sinks"><a class="docs-heading-anchor" href="#Sinks">Sinks</a><a id="Sinks-1"></a><a class="docs-heading-anchor-permalink" href="#Sinks" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="BioRecordsProcessing.Writer" href="#BioRecordsProcessing.Writer"><code>BioRecordsProcessing.Writer</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">Writer(record_module::Module, output_directory::String; suffix = &quot;&quot;)</code></pre><p>Write the output of the processing function into a file, the first argument is the module that owns the <code>Record</code> type (e.g <code>FASTX.FASTA</code>, <code>VCF</code>, ...), and the second the ouput directory. The filename is determined by the source, to which an optional suffix can be added. To avoid overwriting existing files, the pipeline will check that the output file is different from the input file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jonathanBieler/BioRecordsProcessing.jl/blob/03dd05429a737063e83a6e8c225d3c64c19526ed/src/Sink.jl#L42-L51">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BioRecordsProcessing.Collect" href="#BioRecordsProcessing.Collect"><code>BioRecordsProcessing.Collect</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">Collect(T::DataType; paired=false)</code></pre><p>Write the output of the processing function into an vector in memory. The type of output has to be provided. For paired files the option paired need to be set to <code>true</code>, the output will then consists of a vector of tuples.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/jonathanBieler/BioRecordsProcessing.jl/blob/03dd05429a737063e83a6e8c225d3c64c19526ed/src/Sink.jl#L10-L17">source</a></section></article><h3 id="Examples"><a class="docs-heading-anchor" href="#Examples">Examples</a><a id="Examples-1"></a><a class="docs-heading-anchor-permalink" href="#Examples" title="Permalink"></a></h3></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="examples/">Examples »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Wednesday 19 April 2023 07:06">Wednesday 19 April 2023</span>. Using Julia version 1.7.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
